server/proxy_protocol.go

```diff
@@ -0,0 +1,218 @@
+	allowAll := false
+	allowedNets := []*net.IPNet{}
+	if allowedIPs == "*" {
+		allowAll = true
+	} else {
+		for _, aip := range strings.Split(allowedIPs, ",") {
+			saip := strings.TrimSpace(aip)
+			_, ipnet, err := net.ParseCIDR(saip)
+			if err == nil {
+				allowedNets = append(allowedNets, ipnet)
+				continue
+			}
+			psaip := fmt.Sprintf("%s/32", saip)
+			_, ipnet, err = net.ParseCIDR(psaip)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			allowedNets = append(allowedNets, ipnet)
+		}
+	}
+	return &proxyProtocolConnBuilder{
+		allowAll:          allowAll,
+		allowedNets:       allowedNets,
+		headerReadTimeout: headerReadTimeout,
+	}, nil
+}
+
+func (b *proxyProtocolConnBuilder) wrapConn(conn net.Conn) (*proxyProtocolConn, error) {
+	ppconn := &proxyProtocolConn{
+		Conn:    conn,
+		builder: b,
+	}
+	err := ppconn.readClientAddrBehindProxy()
+	return ppconn, errors.Trace(err)
+}
+
+func (b *proxyProtocolConnBuilder) checkAllowed(raddr net.Addr) bool {
+	if b.allowAll {
+		return true
+	}
+	taddr, ok := raddr.(*net.TCPAddr)
+	if !ok {
+		return false
+	}
+	cip := taddr.IP
+	for _, ipnet := range b.allowedNets {
+		if ipnet.Contains(cip) {
+			return true
+		}
+	}
+	return false
+}
+
+type proxyProtocolConn struct {
+	net.Conn
+	builder            *proxyProtocolConnBuilder
+	clientIP           net.Addr
+	exceedBuffer       []byte
+	exceedBufferStart  int
+	exceedBufferLen    int
+	exceedBufferReaded bool
+}
+
+func (c *proxyProtocolConn) readClientAddrBehindProxy() error {
+	connRemoteAddr := c.Conn.RemoteAddr()
+	allowed := c.builder.checkAllowed(connRemoteAddr)
+	if !allowed {
+		return errProxyAddressNotAllowed
+	}
+	return c.parseHeaderV1(connRemoteAddr)
+}
+
+func (c *proxyProtocolConn) parseHeaderV1(connRemoteAddr net.Addr) error {
+	buffer, err := c.readHeaderV1()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	raddr, err := c.extractClientIPV1(buffer, connRemoteAddr)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	c.clientIP = raddr
+	return nil
+}
+
+func (c *proxyProtocolConn) extractClientIPV1(buffer []byte, connRemoteAddr net.Addr) (net.Addr, error) {
+	header := string(buffer)
+	parts := strings.Split(header, " ")
+	if len(parts) != 6 {
+		if len(parts) > 1 && parts[1] == "UNKNOWN\r\n" {
+			return connRemoteAddr, nil
+		}
+		return nil, errProxyProtocolV1HeaderInvalid
+	}
+	clientIPStr := parts[2]
+	clientPortStr := parts[4]
+	iptype := parts[1]
+	switch iptype {
+	case "TCP4":
+		addrStr := fmt.Sprintf("%s:%s", clientIPStr, clientPortStr)
+		return net.ResolveTCPAddr("tcp4", addrStr)
+	case "TCP6":
+		addrStr := fmt.Sprintf("[%s]:%s", clientIPStr, clientPortStr)
+		return net.ResolveTCPAddr("tcp6", addrStr)
+	case "UNKNOWN":
+		return connRemoteAddr, nil
+	default:
+		return nil, errProxyProtocolV1HeaderInvalid
+	}
+}
+
+func (c *proxyProtocolConn) RemoteAddr() net.Addr {
+	return c.clientIP
+}
+
+func (c *proxyProtocolConn) Read(buffer []byte) (int, error) {
```

Comment:
If we use ReadBytes(byte(0x10)) it only work as PROXO protocol v1, if we support v2 in future, ReadBytes call is not fit for this.